conda_build/index/__init__.py

```diff
@@ -921,25 +1089,305 @@ def index_subdir(self, subdir, index_file=None, verbose=False, progress=False):
             #     not using md5 here because it takes too long. If needing to do full md5 checks,
             #     use the --deep-integrity-check flag / self.deep_integrity_check option.
             update_set = self._calculate_update_set(
-                subdir, fns_in_subdir, old_repodata_fns, stat_cache,
-                verbose=verbose, progress=progress
+                subdir,
+                fns_in_subdir,
+                old_repodata_fns,
+                stat_cache,
+                verbose=verbose,
+                progress=progress,
+            )
+            # unchanged_set: packages in old repodata whose information can carry straight
+            #     across to new repodata
+            unchanged_set = old_repodata_fns - update_set - remove_set - ignore_set
+
+            # clean up removed files
+            removed_set = old_repodata_fns - fns_in_subdir
+            for fn in removed_set:
+                if fn in stat_cache:
+                    del stat_cache[fn]
+
+            new_repodata_packages = {
+                k: v
+                for k, v in old_repodata.get("packages", {}).items()
+                if k in unchanged_set
+            }
+            new_repodata_conda_packages = {
+                k: v
+                for k, v in old_repodata.get("packages.conda", {}).items()
+                if k in unchanged_set
+            }
+
+            for k in sorted(unchanged_set):
+                if not (k in new_repodata_packages or k in new_repodata_conda_packages):
+                    # XXX select more than one index at a time
+                    fn, rec = cache.load_index_from_cache(fn)
+                    # this is how we pass an exception through.  When fn == rec, there's been a problem,
+                    #    and we need to reload this file
+                    if fn == rec:
+                        update_set.add(fn)
+                    else:
+                        if fn.endswith(CONDA_PACKAGE_EXTENSION_V1):
+                            new_repodata_packages[fn] = rec
+                        else:
+                            new_repodata_conda_packages[fn] = rec
+
+            # Invalidate cached files for update_set.
+            # Extract and cache update_set and add_set, then add to new_repodata_packages.
+            # This is also where we update the contents of the stat_cache for successfully
+            #   extracted packages.
+            # Sorting here prioritizes .conda files ('c') over .tar.bz2 files ('b')
+            hash_extract_set = tuple(concatv(add_set, update_set))
+
+            extract_func = functools.partial(
+                cache.extract_to_cache, self.channel_root, subdir
```

Comment:
we need a way to preview which files extract_to_cache actually has to open (those not already in the cache), maybe just a function that loads from the cache, and a second one that updates from cache, instead of "load or update as needed"